### ISSUE TITLE
cpu_get_clock: Assertion `increment > 0' failed.

### ISSUE THREAD

**Issue by willsheil** [S2E/s2e-env#170]
I followed the instruction to finish the Coreutils Coverage Testing on the ubuntu 16.04 64bit, but found a problem, when I used debian x86 image, below error showed after running several mins, 

54 [State 413] LinuxMonitor: mmap pid=0x50f addr=0xb77b2000 size=0x3000 prot=0x3 flag=0x12 pgoff=0x1b0
54 [State 413] LinuxMonitor: munmap pid=0x50f start=0xb77b5000 end=0xb77b8000
**qemu-system-i386: /home/will/s2e_new/source/s2e/qemu/cpus.c:156: cpu_get_clock: Assertion `increment > 0' failed.**
./launch-s2e.sh: line 74: 125487 Aborted                 (core dumped) LD_PRELOAD=$LIBS2E $QEMU $DRIVE -k en-us $GRAPHICS -monitor null -m 256M -enable-kvm -serial file:serial.txt -net none -net nic,model=e1000 -loadvm ready $*
Then I changed the image to debian x86-64, showed same error, which happened immediately once after the ./launch-s2e.sh. running. 

Could you pls help to check this issue?Thanks.

**Comment by vitalych** [comment]
Something is wrong here: https://github.com/S2E/qemu/blob/v1.0.50-se/cpus.c#L156
Normally the clock is monotonic, so the assertion must always hold. So either the host's clock is broken or there is a problem with the clock offset.

Is the problem deterministic? Could you reproduce by building S2E in debug mode? If so, please send the core dump and the libs2e.so + qemu binaries.

**Comment by willsheil** [comment]
Yes, it's deterministic.The difference is that some times shows immediately, some times the error shows after running several mins.
I run it on S2E  virtual machine ，I am not sure if this is reason.
I will compile a Debug version, then send related file to you for checking.
Thanks for your help in advance.

**Comment by vitalych** [comment]
Thanks! Maybe it's workload-dependent. Could you also do ``s2e export_project`` so that I could try with the input you have?

**Comment by 0xcpu** [comment]
I have encountered the same issue on the latest `s2e` release. I am trying to build a `Windows 10` image:
`/home/cpu/Tools/s2e/install/bin/qemu-system-x86_64 -no-reboot  -enable-kvm -m 2G -net none -net nic,model=e1000 -cpu core2duo -drive if=ide,index=0,file=/home/cpu/Tools/s2e/images/windows-10ent1703-x86_64/image.raw.tmp,format=raw,cache=writeback -drive if=ide,index=2,file=/home/cpu/Tools/s2e/images/.tmp-output/windows-10ent1703-x86_64/boot.iso,media=cdrom -drive if=ide,index=3,file=/home/cpu/Tools/s2e/images/.tmp-output/windows-10ent1703-x86_64/post_install_files.iso,media=cdrom -serial file:/home/cpu/Tools/s2e/images/windows-10ent1703-x86_64/post_install_serial1.txt -periodic-screenshot /home/cpu/Tools/s2e/images/.tmp-output/windows-10ent1703-x86_64/post_install_screenshots`

`s2e` output:
```
qemu-system-x86_64: /home/cpu/Tools/s2e/source/s2e/qemu/cpus.c:352: cpu_get_clock_locked: Assertion `increment > 0' failed.
/home/cpu/Tools/s2e/source/guest-images//Makefile.windows:353: recipe for target '/home/cpu/Tools/s2e/images/windows-10ent1703-x86_64/image.raw.s2e' failed
make: *** [/home/cpu/Tools/s2e/images/windows-10ent1703-x86_64/image.raw.s2e] Aborted (core dumped)
```

My environment, an `Ubuntu VM` running on top of `Hyper-V`:
```
LSB Version:	core-9.20170808ubuntu1-noarch:printing-9.20170808ubuntu1-noarch:security-9.20170808ubuntu1-noarch
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic
```

**Comment by vitalych** [comment]
Thanks for reporting. HyperV is something new, I never tried it. I tested on VMware and VirtualBox so far.
A few ideas to investigate:
- Is it deterministic? Copy/paste the qemu command line above in the terminal and run it again.
- If it still crashes, could you comment out the assert? It might let you through with the installation.
- You could also check with upstream QEMU 1.0, just to double check that it works properly. At least this would rule out a QEMU bug. Just install manually with the command you pasted above (remove the -periodic-screenshot command).

**Comment by 0xcpu** [comment]
> 
> 
> Thanks for reporting. HyperV is something new, I never tried it. I tested on VMware and VirtualBox so far.
> A few ideas to investigate:
> 
>     * Is it deterministic? Copy/paste the qemu command line above in the terminal and run it again.
> 
>     * If it still crashes, could you comment out the assert? It might let you through with the installation.
> 
>     * You could also check with upstream QEMU 1.0, just to double check that it works properly. At least this would rule out a QEMU bug. Just install manually with the command you pasted above (remove the -periodic-screenshot command).

- Yes, it is deterministic, already tried the copy/paste method.
- Will check.
- Will check.

Thank you!

**Comment by vitalych** [comment]
It's great that it's deterministic. Does it happen shortly after that you start QEMU or long into the installation? Also try with the debug build of QEMU (build/s2e/qemu-debug/...). If it reproduces with the debug build, please attach the link to the binary and the core dump so that I can have a look.

**Comment by vitalych** [comment]
I'd be curious to see the content of the `increment` variable. If it's 0, means that the host clock did not advance since the last call. QEMU uses a monotonic one, so it must not happen.

Also check if `dmesg` on your host has any suspicious things.

**Comment by 0xcpu** [comment]
> 
> 
> It's great that it's deterministic. Does it happen shortly after that you start QEMU or long into the installation? Also try with the debug build of QEMU (build/s2e/qemu-debug/...). If it reproduces with the debug build, please attach the link to the binary and the core dump so that I can have a look.

Commented the assertion, recompiled. Seems to work, currently installing the OS.

**Comment by 0xcpu** [comment]
> 
> 
> I'd be curious to see the content of the `increment` variable. If it's 0, means that the host clock did not advance since the last call. QEMU uses a monotonic one, so it must not happen.
> 
> Also check if `dmesg` on your host has any suspicious things.

I can tell that only after I recompile `qemu` again...

**Comment by 0xcpu** [comment]
Recompiled `qemu` to log the value of `increment`:
```
cur_clock(65967919800) - cpu_clock_prev(65967584400) = increment(335400)
cur_clock(65967950600) - cpu_clock_prev(65967919800) = increment(30800)
cur_clock(65968093700) - cpu_clock_prev(65967950600) = increment(143100)
cur_clock(66177926500) - cpu_clock_prev(65968093700) = increment(209832800)
...
```
As you can deduce, the assertion doesn't trigger anymore and that is after I installed the OS. So, not sure what can be the cause of the initial issue, but recompiling `qemu` without that assertion solved the issue, later it possible to reactivate the assertion if one wills.

**Comment by likuilin** [comment]
I ran into this problem during image building on HyperV on a Windows host, and `/btrfs/s2e/s2e-dir/install/bin/qemu-system-x86_64 -nographic` reproduces it for me.

Interestingly, the problem goes away if qemu is run under gdb with a breakpoint and auto-continue (via `command`) before the `clock_gettime` call. So, this is timing-based, and the host is executing too quickly - all signs point to this just being the clock not advancing between two measurements. 

The `assert(increment > 0)` was added by S2E, actually, here: https://github.com/S2E/qemu/commit/f50d282fed3380303212923f3e9acc82061ba5ed#diff-9f950c50f1249c4a38ec8f8f3814feeda37d6228c4fa43d7c9d9b61909d82279R352

> I'd be curious to see the content of the `increment` variable. If it's 0, means that the host clock did not advance since the last call. QEMU uses a monotonic one, so it must not happen.

@vitalych Monotonic just means it can never decrease, not that every successive call to clock_gettime for the monotonic clock must return a higher value. So I think we should fix this bug by just removing the assert - it doesn't seem to affect the clock scaling feature if `increment` happens to be zero there, too.
